docs/rfcs/GPUXDialect.md

```diff
@@ -0,0 +1,226 @@
+
+
+#### gpux.destroy_stream (gpux::DestroyStreamOp)
+
+Destroy the GPU Stream.
+
+Syntax:
+
+```
+operation ::= "gpux.destroy_stream"() : () -> !gpux.StreamType
+```
+
+The gpux.destroy_stream() operation will be added by a custom pass only if the stream is not provided by the user. If the user provides the stream, we expect the user to manage the lifetime of the stream.
+If user does not provide the stream, this operation will destroy the default stream.
+
+
+#### gpux.alloc (gpux::AllocOp)
+
+GPU memory allocation operation.
+
+Syntax:
+
+```
+operation ::= gpux.alloc custom<AsyncDependencies>(type($asyncToken), $asyncDependencies)
+              ($gpux_stream, $dynamicSizes) ([ $symbolOperands^ ])? attr-dict : type($memref)
+```
+
+This op is an extension of upstream gpu.alloc op with one added argument for stream. Stream here will be an optional argument, so when upstreaming the code, the upstream GPU dialect users have no change in their usage of this op.
+
+
+#### gpux.dealloc (gpux::DeallocOp)
+
+GPU memory deallocation operation.
+
+Syntax:
+
+```
+operation ::= gpux.dealloc custom<AsyncDependencies>(type($asyncToken), $asyncDependencies)
+              ($gpux_stream, $memref) attr-dict : type($memref)
+```
+
+This op is an extension of upstream gpu.dealloc op with one added argument for stream. Stream here will be an optional argument, so when upstreaming the code, the upstream GPU dialect users have no change in their usage of this op.
+
+
+#### gpux.launch_func (gpux::LaunchFuncOp)
+
+Launches a function as a GPU kernel.
+
+Syntax:
+
+```
+operation ::= gpux.launch_func custom<AsyncDependencies>(type($asyncToken), $asyncDependencies)
+              $kernel
+              blocks in  ($gridSizeX, $gridSizeY, $gridSizeZ)
+              threads in ($blockSizeX, $blockSizeY, $blockSizeZ)
+              (dynamic_shared_memory_size $dynamicSharedMemorySize^)?
+              custom<LaunchFuncOperands>($gpux_stream, $operands, type($operands)) attr-dict
+```
+
+This op is an extension of upstream gpu.launch_func op with one added argument for stream. Stream here will be an optional argument, so when upstreaming the code, the upstream GPU dialect users have no change in their usage of this op.
+
+#### gpux.wait (gpux::WaitOp)
+
+Wait for gpu ops in a particular stream to complete.
+
+Syntax:
+
+```
+operation ::= gpu.wait custom<AsyncDependencies>(type($asyncToken), $asyncDependencies)
+              $gpux_stream attr-dict
+```
+
+The wait op will need to be provided with the stream to wait on. This op is an extension of upstream gpu.wait op with one added argument for stream. Stream here will be an optional argument, so when upstreaming the code, the upstream GPU dialect users have no change in their usage of this op.
+
+## Example Usage in case of user provided stream:
+
+// IR before our custom transformation pass:
+
+```
+
+func.func @main(%stream : !gpux.StreamType) attributes {llvm.emit_c_interface} {
```

Comment:
This IR doesn't make sense. When user transforms to GPU dialect, user either choose to not specify stream. If user choose to user stream, it is user's responsibility to use the stream to GPU ops.

Reply:
>>>it is user's responsibility to use the stream to GPU ops.

how will the user pass the stream to gpu ops ? they have to use this GPUX dialect to pass the stream to GPU ops...are you saying user will have their own pass to convert to the GPUX dialect which adds stream as an input to the upstream gpu ops?